### ISSUE TITLE
param-out not taken into account when using generics

### ISSUE THREAD

**Issue by greg0ire** [vimeo/psalm#8891]
Psalm doesn't seem to realize that 42 is an integer here:

https://psalm.dev/r/654392acdb

It is suggesting to use `param-out`, which is confusing as I'm already using it.

**Comment by weirdan** [comment]
The snippet seems outdated and does not illustrate the issue.

**Comment by greg0ire** [comment]
You're right, I made a copy-paste mistake when examining another issue :sweat_smile: Thanks for pointing that out! Here is what I meant to post: https://psalm.dev/r/e46c3120a6

**Comment by weirdan** [comment]
While `42` is int, it's not necessarily `T`. E.g. when you have `X<13>`, `T` is `13`, but you're outputting `42`.

**Comment by weirdan** [comment]
No, that doesn't make much sense to me (of course, it should be `int &$s`).

**Comment by greg0ire** [comment]
Should it? If the param enters the method as `'42'`, and then is set to `42`, then this should be correct right (and is the purpose of `param-out`? Here an example: https://github.com/doctrine/lexer/blob/2.0.x/tests/ConcreteLexer.php (not sure how realistic it is though, I known `doctrine/orm` doesn't do that kind of stuff).

**Comment by greg0ire** [comment]
Here is an example with just Psalm: https://psalm.dev/r/b7de7cfb93
And that same code running: https://3v4l.org/jUfG8
